xds/internal/server/rds_handler.go

```diff
@@ -0,0 +1,137 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package server
+
+import (
+	"sync"
+
+	"google.golang.org/grpc/xds/internal/xdsclient"
+)
+
+// rdsHandlerUpdate wraps the full RouteConfigUpdate that are dynamically queried for a given
+// server side listener.
+type rdsHandlerUpdate struct {
+	updates map[string]xdsclient.RouteConfigUpdate
+	err     error
+}
+
+// rdsHandler handles any RDS queries that need to be started for a given server
+// side listeners Filter Chains (i.e. not inline).
+type rdsHandler struct {
+	parent *listenerWrapper
+
+	mu sync.Mutex
+
+	rdsUpdates map[string]xdsclient.RouteConfigUpdate
+	rdsCancels map[string]func()
+
+	// For a rdsHandler update, the only update wrapped listener cares about is
+	// most recent one, so this channel will be opportunistically drained before
+	// sending any new updates.
+	updateChannel chan rdsHandlerUpdate
+}
+
+// newRDSHandler is expected to called once on instantiation of a wrapped
+// listener. On any LDS updates the wrapped listener receives, the listener
+// should update the handler with the route names (which specify dynamic RDS)
+// using the function below.
+func newRDSHandler(parent *listenerWrapper) *rdsHandler {
+	return &rdsHandler{
+		parent:        parent,
+		updateChannel: make(chan rdsHandlerUpdate, 1),
+		rdsUpdates:    make(map[string]xdsclient.RouteConfigUpdate),
+		rdsCancels:    make(map[string]func()),
+	}
+}
+
+// updateRouteNamesToWatch handles a list of route names to watch for a given
+// server side listener (if a filter chain specifies dynamic RDS configuration).
+// This function handles all the logic with respect to any routes that may have
+// been added or deleted as compared to what was previously present.
+func (rh *rdsHandler) updateRouteNamesToWatch(routeNamesToWatch map[string]bool) {
+	rh.mu.Lock()
+	defer rh.mu.Unlock()
+	// Add and start watches for any routes for any new routes in routeNamesToWatch.
+	for routeName := range routeNamesToWatch {
+		if _, ok := rh.rdsCancels[routeName]; !ok {
+			func(routeName string) {
+				rh.rdsCancels[routeName] = rh.parent.xdsC.WatchRouteConfig(routeName, func(update xdsclient.RouteConfigUpdate, err error) {
+					rh.handleRouteUpdate(routeName, update, err)
+				})
+			}(routeName)
+		}
+	}
+
+	// Delete and cancel watches for any routes from persisted routeNamesToWatch
+	// that are no longer present.
+	for routeName := range rh.rdsCancels {
+		if _, ok := routeNamesToWatch[routeName]; !ok {
+			rh.rdsCancels[routeName]()
+			delete(rh.rdsCancels, routeName)
+			delete(rh.rdsUpdates, routeName)
+		}
+	}
+
+	// If the full list (determined by length) of updates are now successfully
+	// updated, the listener is ready to be updated.
+	if len(rh.rdsUpdates) == len(rh.rdsCancels) && len(routeNamesToWatch) != 0 {
+		select {
+		case <-rh.updateChannel:
+		default:
+		}
+		rh.updateChannel <- rdsHandlerUpdate{updates: rh.rdsUpdates}
```

Comment:
This pattern is repeated three times in very close succession and in a reasonably small chunk of code. You could make a helper function.

```golang
func drainAndPush(ch chan rdsHandlerUpdate, update rdsHandlerUpdate) {
	select {
	case <-ch:
	default:
	}
	ch <- update
}
```

Reply:
Done. Thanks for the idea :D!